### ISSUE TITLE
Support deployment on Sepolia

### ISSUE THREAD

**Issue by michalinacienciala** [threshold-network/solidity-contracts#150]
### Why?
The Görli testnet currently used by Threshold/Keep for development purposes is planned to become deprecated with the end of year 2023. The testnet that is planned to replace it is called [Holešky](https://github.com/eth-clients/holesky), however it's not yet available - it's planned it will become widely accessible on Oct 1, 2023 ([source](https://everstake.one/blog/new-ethereum-testnet-holesky-all-you-need-to-know-now)).
Switching our infrastructure to support new testnet is quite time consuming, so moving directly from Görli to Holešky may be quite risky, especially if there would be some delays in the date of Holešky genesis (not meeting the planned timelines is not a rare occurrence in the Ethereum space). As a solution, we decided to switch first to another testnet that is currently live - Sepolia. This testnet's EOL is planned for 2026, which gives us plenty of time to move to Holešky before Sepolia gets deprecated.

**Testnets Lifecycle:**
![image](https://github.com/threshold-network/solidity-contracts/assets/78352137/fca11ea0-ad73-43e1-af33-e1aed24a545b)
Read more in [Predictable Ethereum Testnet Lifecycle Proposal](https://ethereum-magicians.org/t/proposal-predictable-ethereum-testnet-lifecycle/11575).

### How?
Deploying contracts to a new testnet and adjusting CI accordingly is a work consisting of many tasks and requiring coordination of several team members. We shouldn't get rid of Görli support until we're sure that everything works on Sepolia smoothly (maybe even until Görli gets deprecated). We may consider adding Holešky support together with Sepolia in some places, for the future use (for example in Hardhat config).

### What & who?

- [x] **Update** list of supported networks in **`keep-common`** (https://github.com/keep-network/keep-common/pull/118)
- [ ] Adjust our **Hardhat/Truffle configuration** to support Sepolia (without removing configuration for Görli yet):
  - [x] `@keep-network/keep-core` - https://github.com/keep-network/keep-core/pull/3712
  - [x] `@keep-network/keep-ecdsa` - modified locally (repository deprecated)
  - [x] `@keep-network/tbtc` - modified locally (repository deprecated)
  - [x] `@threshold-network/solidity-contracts` - https://github.com/threshold-network/solidity-contracts/pull/151, https://github.com/threshold-network/solidity-contracts/pull/152, https://github.com/threshold-network/solidity-contracts/pull/154
  - [x] `@keep-network/coverage-pools` - https://github.com/keep-network/coverage-pools/pull/235
  - [x] `@keep-network/random-beacon` - https://github.com/keep-network/keep-core/pull/3706
  - [x] `@keep-network/ecdsa` - https://github.com/keep-network/keep-core/pull/3706
  - [x] `@keep-network/tbtc-v2` - https://github.com/keep-network/tbtc-v2/pull/691
  - [x] `@keep-network/tbtc-v2.ts` - https://github.com/keep-network/tbtc-v2/pull/691
  - [ ] `@keep-network/tbtc-v2-arbitrum` (<ins>nice to have</ins>) - https://github.com/keep-network/tbtc-v2/pull/753
  - [ ] `@keep-network/tbtc-v2-optimism` (<ins>nice to have</ins>) - https://github.com/keep-network/tbtc-v2/pull/752
  - [ ] `@keep-network/tbtc-v2-polygon` (<ins>nice to have</ins>) - https://github.com/keep-network/tbtc-v2/pull/751
  - [ ] `@keep-network/tbtc-v2-solana`? (once Solana support lands on `main`) (<ins>nice to have</ins>)
  - [ ] `@keep-network/tbtc-v2-base` - https://github.com/keep-network/tbtc-v2/pull/742
- [x] **Create `SepoliaLightRelay.sol`** - https://github.com/keep-network/tbtc-v2/pull/691
- [x] **Adjust deployments scripts**, where needed.
- [x] Send SepoliaETH to the account that will be used as a deployer (`0x68ad60CC5e8f3B7cC53beaB321cf0e6036962dBc`)
- [ ] **Deploy contracts** on Sepolia and **publish packages** with them to the NPM registry (in specific order, taking under consideration dependencies between the projects):
  - [x] `@keep-network/keep-core@1.8.1-sepolia.0`
  - [x] `@keep-network/keep-ecdsa@1.9.0-sepolia.0`
  - [x] `@keep-network/tbtc@1.1.2-sepolia.0`
  - [x] `@threshold-network/solidity-contracts@1.3.0-sepolia.0`
  - [x] `@keep-network/coverage-pools@2.1.0-sepolia.0`
  - [ ] `@keep-network/coverage-pools@1...` (There were problems with publishing package for v1, we will skip it's deployment and manage without it, at least for now)
  - [x] `@keep-network/random-beacon@2.1.0-sepolia.1`
  - [x] `@keep-network/ecdsa@2.1.0-sepolia.1`
  - [x] `@keep-network/tbtc-v2@1.6.0-sepolia.0`
  - [x] `keep-network/tbtc-v2.ts@1.3.0-sepolia.0` (published based on code from `19e2dfff` commit, package marked as deprecated)
  - [ ] `@keep-network/tbtc-v2-arbitrum` (also update  `.gitignore` and `package.json`) (<ins>nice to have</ins>)
  - [ ] `@keep-network/tbtc-v2-optimism` (also update `.gitignore` and `package.json`) (<ins>nice to have</ins>)
  - [ ] `@keep-network/tbtc-v2-polygon` (also update `.gitignore` and `package.json`) (<ins>nice to have</ins>)
  - [ ] `@keep-network/tbtc-v2-solana`? (also update `.gitignore` and `package.json`) (once Solana support lands on `main`) (<ins>nice to have</ins>)
  - [ ] `@keep-network/tbtc-v2-base`
- [x] **Verify contracts** on Etherscan
- [x] **Update `merkle-drop` config** - https://github.com/threshold-network/merkle-distribution/pull/93/
- [x] **Deploy merkle drop and stub(?) `SimplePreApplication`** on Sepolia
- [x] **Update CI config file** (`keep-network/ci/config/env/sepolia.env`) - https://github.com/keep-network/ci/pull/48
- [x] **Setup environment variables on organization level (in `keep-network` and `threshold-network`)**:
  - [x] `SEPOLIA_ETH_HOSTNAME_HTTP`
  - [x] `SEPOLIA_ETH_HOSTNAME_WS`
  - [x] `TESTNET_ETH_CONTRACT_OWNER_PRIVATE_KEY`
  - [x] `DAPP_DEV_TESTNET_ETH_CONTRACT_OWNER_PRIVATE_KEY`
- [x] **Modify CI `.yml` files** in following projects to use above secrets when deploying on Sepolia:
  - [x] `threshold-network/solidity-contracts`
  - [x] `keep-network/coverage-pools` (?)
  - [x] `keep-network/random-beacon`
  - [x] `keep-network/ecdsa`
  - [x] `keep-network/tbtc-v2`
  - [x] `keep-network/tbtc-v2/monitoring` - https://github.com/keep-network/tbtc-v2/pull/754
  - [x] `threshold-network/token-dashboard` - https://github.com/threshold-network/token-dashboard/pull/605
- [ ] (<ins>Nice to have</ins>) **Include publishing of** following **packages in CI flow**:
  - [ ] `keep-network/tbtc-v2-arbitrum` (also update `.gitignore` and `package.json`)
  - [ ] `keep-network/tbtc-v2-optimism` (also update `.gitignore` and `package.json`)
  - [ ] `keep-network/tbtc-v2-polygon` (also update `.gitignore` and `package.json`)
  - [ ] `keep-network/tbtc-v2-solana`? (also update `.gitignore` and `package.json`) (once Solana support lands on `main`)
  - [ ] `keep-network/tbtc-v2-base` (also update `.gitignore` and `package.json`)
- [x] **Modify `dapp-development` branch** in following projects to support Sepolia:
  - [x] `threshold-network/solidity-contracts` - https://github.com/threshold-network/solidity-contracts/pull/118
  - [x] `keep-network/keep-core` - https://github.com/keep-network/keep-core/pull/3118
  - [x] `keep-network/tbtc-v2` - https://github.com/keep-network/tbtc-v2/pull/403
- [x] **Create `dapp-development-sepolia`-tagged packages** for:
  - [x] `@threshold-network/solidity-contracts@1.3.0-dapp-dev-sepolia.0`
  - [x] `@keep-network/random-beacon@2.1.0-dapp-dev-sepolia.0`
  - [x] `@keep-network/ecdsa@2.1.0-dapp-dev-sepolia.0`
  - [x] `@keep-network/tbtc-v2@1.6.0-dapp-dev-sepolia.0`
- [x] **Document** the published contracts in the Threshold docs and update READMEs and other documentation referencing `goerli` - https://app.gitbook.com/o/R2meumXNNad4y1B10iL7/s/WosjlL4zUGUMlcMfuSAp/~/changes/268/resources/contract-addresses/sepolia-testnet, https://app.gitbook.com/o/R2meumXNNad4y1B10iL7/s/WosjlL4zUGUMlcMfuSAp/~/changes/326/app-development/tbtc-v2/tbtc-sdk/guides/initialize-sdk
- [x] **Update subgraphs and TBTC testnet scan website** (https://testnet.tbtcscan.com/) to point to Sepolia (or allow to switch between Görli and Sepolia) - ([repo link](https://github.com/suntzu93/tbtcv2_website_info))
- [x] Adjust governable parameters in contracts to make testing easier:
  - [x] Update wallet creation parameters to allow creating wallets one after another without delay
  - [x] Updatee DKG parameters to reduce approval time from 48 hours to 10 minutes and reduce the timeout to quickly terminate DKGs in case of failures
  - [x] Reduce optimistic minting delay to 1 hour (needs to be finalized)
  - [x] Authorize our minter and guardian
  - [x] Configure SepoliaLightRelay to accept testnet SPV proofs
  - [x] Fund the reimbursement pool with 5 ETH
  - [x] Fund maintainer account with 1 ETH
  - [x] Fund minter and guardian with 1 ETH
- [x] Deploy goerli-based dApp under a separate domain, e.g. dashboard-goerli.test.threshold.network
- [x] **Update mentions of `GoerliLightRelay`** in `keep-core` to mention also SepoliaLightRelay - https://github.com/keep-network/keep-core/pull/3730
- [x] Update client config, apply - https://github.com/keep-network/keep-core/pull/3741
- [x] Update maintainer config, apply (https://github.com/keep-network/keep-core/pull/3742)
- [x] **Update `optimistic-minting`** configuration (https://github.com/keep-network/optimistic-minting/pull/101)
- [x] Update monitoring (https://github.com/keep-network/tbtc-v2/pull/754)
- [x] **Adjust dApp** to use Sepolia-based contracts (similarly to https://github.com/threshold-network/token-dashboard/pull/138) - https://github.com/threshold-network/token-dashboard/pull/605
- [x] **Distribute SepoliaETH** (or inform about the faucet) to devs that need it
- [ ] Perform **cleanup** once we confirm everything works smoothly and Görli gets deprecated:
  - [ ] remove the Hardhat, CI configuration relating to Görli testnet:
    - [ ] https://github.com/keep-network/keep-core/pull/3746
    - [ ] https://github.com/threshold-network/solidity-contracts/pull/160
    - [ ] https://github.com/keep-network/keep-common/pull/119
    - [ ] https://github.com/keep-network/tbtc-v2/pull/755
    - [ ] https://github.com/threshold-network/token-dashboard/pull/670
    - [x] https://github.com/threshold-network/merkle-distribution/pull/102
  - [ ] remove the Görli-related GH secrets
  - [ ] remove mentions of Görli in documentation
  - [ ] deprecate `goerli`-related NPM packages (we can also do a cleanup of other obsolete packages)
  - [ ] get rid of the goerli-based testnet dApp and domain
- [ ] Adjust `optimistic-minting`'s integration tests for Sepolia contracts

**Comment by michalinacienciala** [comment]
Documenting some decisions and steps of the process:

**Deployer account**
We decided to deploy the non-dapp-dev contracts on Sepolia using the `0x68ad60CC5e8f3B7cC53beaB321cf0e6036962dBc` account. This is an account that was also used for Goerli deployment. We'll use it to deploy both Keep (v1, v2) and Threshold contracts.

**Deploying legacy v1 Keep contracts**
As some of the legacy projects are deprecated, we didn't commit the changes needed for Sepolia deployment to the repos. Only did this for `keep-core` project.
Here is the list of updates that was done in the deprecated projects to publish the NPM packages:
- `keep-core`: changes in code as in https://github.com/keep-network/keep-core/pull/3712, then:
```
nvm use 16
npm ci
CHAIN_API_URL=... CONTRACT_OWNER_ACCOUNT_PRIVATE_KEY=... npx truffle migrate --reset --network sepolia
mkdir -p artifacts
cp -r build/contracts/* artifacts/
```
then updated the package version in the `package.json` to `1.8.1-sepolia.0` and then:
```
npm publish --access=public --dry-run --tag sepolia
npm publish --access=public --tag sepolia          # required running `npm addr` first
```

- `keep-ecdsa`: updated `truffle`-related dependencies using Node.js v16, added `sepolia` config in `solidity/truffle.js` and then in the `solidity` folder:
```
nvm use 16
npm ci
npm install --save-exact @keep-network/keep-core@1.8.1-sepolia.0
CHAIN_API_URL=... CONTRACT_OWNER_ACCOUNT_PRIVATE_KEY=... npx truffle migrate --reset --network sepolia
mkdir -p artifacts
cp -r build/contracts/* artifacts/
```
then updated the package version in the `package.json` to `1.9.0-sepolia.0` and then:
```
npm publish --access=public --dry-run --tag sepolia
npm publish --access=public --tag sepolia
```
